### PR TITLE
[core] Fix WiFi connection in safe mode

### LINKED ISSUE #16262: safe_mode not working
### The problem

When entering safe_mode, wifi does not come up, making it useless.

### Which version of ESPHome has the issue?

2026.5.0-dev

### What type of installation are you using?

pip

### What platform are you using?

RTL87XX

### Component causing the issue

safe_mode

### YAML Config

```yaml
esphome:
  name: bw15-1

rtl87xx:
  board: generic-rtl8720cf-2mb-992k

# Enable logging
logger:

# Enable Home Assistant API
api:

ota:
  - platform: esphome
    password: ""

wifi:
  ssid: !secret wifi2_ssid
  password: !secret wifi2_password

captive_portal:
```

### Anything in the logs that might be useful for us?

```txt
Log show that all components are initialized, but afterwards loop is not entered:

Initializing WIFI ...
WIFI initialized
[D][wifi:1298]: Starting scan
[C][component:246]: Setup wifi took 166ms
[I][app:113]: setup() finished successfully!
[W][safe_mode:164]: SAFE MODE IS ACTIVE
[I][app:144]: ESPHome version 2026.5.0-dev compiled on 2026-05-05 21:03:28 +0200
[C][logger:219]: Logger:
  Max Level: DEBUG
  Initial Level: DEBUG
[C][logger:226]:   Log Baud Rate: 115200
  Hardware UART: UART2
[C][logger:235]:   Task Log Buffer Size: 768 bytes
[C][captive_portal:134]: Captive Portal:
[C][wifi:1526]: WiFi:
  Local MAC: 94:C9:60:3B:99:8F
  Connected: NO
[C][esphome.ota:097]: Over-The-Air updates:
  Address: bw15-1.local:8892
  Version: 2
[C][safe_mode:026]: Safe Mode:
  Successful after: 60s
  Invoke after: 10 attempts
  Duration: 300s
[W][safe_mode:051]: SAFE MODE IS ACTIVE
[C][mdns:230]: mDNS:
  Hostname: bw15-1
[D][preferences:117]: Writing 1 items: 0 cached, 1 written, 0 failed

(no more log entries, not connecting to access point, but after 300 seconds safe_mode ends as expected)
```

### Additional information

When looking at the generated main.cpp file,

App.looping_components_.init(ESPHOME_LOOPING_COMPONENT_COUNT);

is not called because

if (safe_mode_safemodecomponent_id->should_enter_safe_mode(10, 300000, 60000)) return;

returns earlier.

After botching a

if (!this->looping_components_.capacity()) this->looping_components_.init(10);

into Application::setup(), startup in safe_mode is successful.

So it seems to be a missing init of Application::looping_components_

### PR DESCRIPTION
**Pull request by Mat931**
# What does this implement/fix?

Fixes a code-generation ordering bug that prevented WiFi from connecting (and the loop from running at all) when `safe_mode` was active.

## Root cause

`safe_mode` runs at `CoroPriority.APPLICATION` (50) and emits

```cpp
if (safe_mode_safemodecomponent_id->should_enter_safe_mode(...)) return;
```

into the generated `main()`. The pre-existing `App.looping_components_.init(ESPHOME_LOOPING_COMPONENT_COUNT)` call lived in `_add_looping_components()` at `CoroPriority.FINAL` (-1000), so it was emitted *after* the safe_mode early return. When safe mode was entered, the `FixedVector<Component *>` storing looping components was never sized, `looping_components_active_end_` stayed at 0, and `loop()` never ran. WiFi therefore never connected and the device stayed stuck in safe mode until timeout.

## Fix

Move the `App.looping_components_.init(...)` emission into core `to_code()` (`CoroPriority.CORE`, 100), so it lands in `main()` before the safe_mode early return. The `static constexpr size_t ESPHOME_LOOPING_COMPONENT_COUNT` declaration stays in `_add_looping_components()` because it is added via `cg.add_global` (file scope), so coroutine ordering does not affect its visibility.

Also handle the corner case where no looping components are registered: a fallback `["0"]` ensures the generated constexpr is `= 0;` instead of an empty (uncompilable) expression.

## Tests

Added regression tests in `tests/unit_tests/core/test_config.py`:

- `test_to_code_priority_above_safe_mode`: guards the ordering invariant that broke #16262.
- `test_add_looping_components_handles_empty_entries`: guards the empty-list constexpr edge case.
- `test_add_looping_components_with_entries`: verifies dedup-by-type and per-type counts in the `HasLoopOverride<T>::value` sum.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected) — [policy](https://developers.esphome.io/contributing/code/#what-constitutes-a-c-breaking-change)
- [ ] Developer breaking change (an API change that could break external components) — [policy](https://developers.esphome.io/contributing/code/#what-is-considered-public-c-api)
- [ ] Undocumented C++ API change (removal or change of undocumented public methods that lambda users may depend on) — [policy](https://developers.esphome.io/contributing/code/#c-user-expectations)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes https://github.com/esphome/esphome/issues/16262

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040/RP2350
- [ ] BK72xx
- [ ] RTL87xx
- [ ] LN882x
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/core/config.py [hunk 1]
```diff
@@ -574,8 +574,6 @@ async def _add_looping_components() -> None:
     # all component types are in scope. calculate_looping_components_() then skips
     # the counting pass and only does the two population passes.
     entries = CORE.data.get("looping_component_entries", [])
-    if not entries:
-        return
```
Comment (bdraco) [inline esphome/core/config.py:578]: this is probably unsafe because they could have 0 looping components

**Comment by bdraco** [comment]
fix looks good otherwise. thanks

File: esphome/core/config.py
Comment (bdraco) [inline esphome/core/config.py]:
```suggestion
    ] or ["0"]
```

**Comment by Mat931** [comment]
Saw your commit too late

**Review by bdraco — DISMISSED** [review]

**Comment by bdraco** [comment]
Traveling so can't do a functional test but issue is strait-forward

**Comment by bdraco** [comment]
Thanks @Mat931

**Comment by bdraco** [comment]
Probably should write an integration test for this as well so we know for sure it never regresses again. let me see if I can get that done before I go to the airport.

**Comment by Mat931** [comment]
OTA updating while in safe mode works, and manually setting `entries = []` compiles too, this is the output in main.cpp

```
static constexpr size_t ESPHOME_LOOPING_COMPONENT_COUNT = \
  0;
```

**Review by bdraco — APPROVED** [review]
